modules/account-lib/src/coin/algo/keyRegistrationBuilder.ts

```diff
@@ -1,26 +1,121 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import BigNumber from 'bignumber.js';
+import algosdk from 'algosdk';
 import { BaseCoin as CoinConfig } from '@bitgo/statics';
-import { NotImplementedError } from '../baseCoin/errors';
-import { BaseKey } from '../baseCoin/iface';
+import { InvalidParameterValueError, InvalidTransactionError } from '../baseCoin/errors';
 import { TransactionBuilder } from './transactionBuilder';
 import { Transaction } from './transaction';
-
+import { TransactionType } from '../baseCoin';
+import { KeyRegTxnSchema } from './txnSchema';
 export class KeyRegistrationBuilder extends TransactionBuilder {
+  protected _voteKey: string;
+  protected _selectionKey: string;
+  protected _voteFirst: number;
+  protected _voteLast: number;
+  protected _voteKeyDilution: number;
+
   constructor(coinConfig: Readonly<CoinConfig>) {
     super(coinConfig);
   }
 
+  voteKey(key: string): KeyRegistrationBuilder {
+    if (!key) {
+      throw new InvalidParameterValueError('voteKey can not be undefined');
+    }
+    this._voteKey = key;
+    return this;
+  }
+
+  selectionKey(key: string): KeyRegistrationBuilder {
+    if (!key) {
+      throw new InvalidParameterValueError('selectionKey can not be undefined');
+    }
+    this._selectionKey = key;
+    return this;
+  }
+
+  voteFirst(round: number): KeyRegistrationBuilder {
+    this.validateValue(new BigNumber(round));
+    this._voteFirst = round;
+    return this;
+  }
+
+  /**
+   *
+   * @param {number} round No theoretical limit. A recommended range is 3,000,000 rounds.
+   * https://developer.algorand.org/docs/run-a-node/participate/generate_keys/
+   */
+  voteLast(round: number = 3000000): KeyRegistrationBuilder {
+    this.validateValue(new BigNumber(round));
+    if (this._voteFirst && round <= this._voteFirst) {
+      throw new Error('VoteKey last round must be greater than first round');
+    }
+    this._voteLast = round;
+    return this;
+  }
+
+  /**
+   * @param {number} size Defaults to 10,000. To reduce the size of the participation key, set the key dilution value to roughly the square root of the range that the partkey is valid for.
+   * https://developer.algorand.org/docs/run-a-node/participate/generate_keys/#generate-the-participation-key-with-goal
+   */
+  voteKeyDilution(size: number = 10000): KeyRegistrationBuilder {
+    this.validateValue(new BigNumber(size));
+    if (this._voteFirst && this._voteLast) {
+      if (size <= Math.sqrt(this._voteLast - this._voteFirst)) {
```

Comment:
this could be the sqrt of a negative number is voteLast is set, then voteFirst

Reply:
Update: voteLast will throw an error if it is being set before voteLast